### ISSUE TITLE
TypeScript type definitions

### ISSUE THREAD

**Issue by knikel** [Automattic/juice#234]
I’ve used `juice` in a recent project, since the project is written in TypeScript, I had to write type definitions for most of `juice`’s APIs. I'm wondering if there is interest in having type definitions as part of the `juice` project?

To do this properly for `typescript@2.*` this would require:
* `juice.d.ts` file
* `test-juice.ts` file with tests which has to compile with correct type definitions taken from `juice.d.ts` file
* `types` entry in `package.json` which points to `juice.d.ts`
* extra `devDependency` on `typescript`

I’m happy to write types for the whole API and make a PR. If there is no interest in that, please close this issue.

**Comment by jcristovao** [comment]
I know I would definitely would be interested in this. If this is not accepted here (which would be preferable), you could otherwise try to submit them to typings or DefinitelyTyped

**Comment by jrit** [comment]
My main concern with this is maintaining it. Do projects usually add something to their tests to check the definitions or is it just a manual process if the API changes? Sorry, I'm not really familiar with typescript so bear with me.

**Comment by knikel** [comment]
@jrit: Thanks for the reply and your time. That is a valid concern of course, I've outlined the trade-offs below.

*tldr;*
> type definitions testing can be an automatic process; when API changes the `juice.d.ts` would have to be updated manually to reflect API changes

The `test-juice.ts` file would be the test file for type definitions. The way it works is as follows:
1. We import the `juice` library into `test-juice.ts`
2. We call each API with correct type annotations and some mock data as per documentation
3. No compiler errors mean that API type definitions are compliant with current API
4. We call `tsc test-juice.ts` as part of the test suite (`tsc` is a `typescript` compiler, has to be included as a `devDependency`)

Few notes on the API changes:
- The `juice.d.ts` file should be only updated when public facing API changes (the API described in the definition file)
- If there are some new APIs, they have to be added to the `juice.d.ts` file manually, there is no way to do that automatically
- If existing API changes, that change also have to be reflected in `juice.d.ts` file manually
- `juice.d.ts` is source of truth about the API to the compiler, the compiler doesn't test the actual code from the library but uses the data from that file

### Next steps
@jrit If you're OK with the trade-offs listed above, I'm happy to proceed with a PR. I am also happy to maintain the definition file when it would need updating.

### Links
1. Example of a type definition test file: [ramda-tests.ts](https://github.com/donnut/typescript-ramda/blob/master/ramda-tests.ts), the `test-juice.ts` would be much simpler of course 😉

**Comment by jrit** [comment]
Sounds good to me. Thanks for explaining it all. There is a CONTRIBUTING.md - can you add a note about the TypeScript requirements in there as well when you prepare the PR?
